JS Unit tests, some proposals and refactoring JS Unit Tests

Amstutz opened the pull request:
I am currently working on a new UI Component involving some js parts, which I would like to unit tests. While implementing the tests according to the [js-unit-test docs](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/development/js/js-unit-test.md) , I observed some aspects, which I believe need some love, before continuing adding new logic.

This PR looks rather large, but no small part is just moving parts around. However, to make this easier to process, here the main parts of it:

- We have currently two folders involving the word *test* on top level, on called *test* the other *tests*. There is no telling from the outside which does what. Also, this further pollutes our already crowded top level. After a short ping with @klees, I hereby propose to have a main [mocha.config.js](https://github.com/ILIAS-eLearning/ILIAS/pull/3685/files#diff-da6caa1bafcadfbad112f260d18d149fb95bc32a59f088e92cfda1398ae83612) file in which we collect all places to look for js tests. The main one being the place, where also the mocha.config.js file is located, tests/UI/Client . The doc is updated accordingly.
- I propose to also test the [example tests](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/development/js/) in the docs directory if npm test is fired, to make sure those updates are kept up-to-date.
- I propose to move from symlinking the files to the test files in this folder, to simply having them stored directly there. As we do with the php unit tests.
- I propose to recursively loop through all folders in tests/UI/Client to make more structure possible. I expect there to be lots of files in future.
- Since in we had already two components in tests/UI/Client already being unit tested, but through the browser and not yet with mocha, I propose hereby to migrate those to mocha testing as well. This made some further changes/additions necessary:
  - I moved both components to the bundling as documented [here](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/development/js/js-bundling.md). This made some slight refactoring necessary.
  - Since some tests really on jQuery and jQuery is used inside the component, I needed to inject jQuery properly as dependency. Also I could not really anymore on il.UI.SomeComponent inside the Components, therefore I needed e.g. to inject counter into NotificationItem. IMO, this is good practice and a nice side effect from unit testing those components through cli with mocha.
  - I needed to load some DOM in the unit tests, to work on with jQuery.
  - To make sure that this DOM stays up to date with DOM rendered by PHP, I added some php tests, explicitly testing, if the DOM loaded in the js Tests is actually DOM we render in php. IMO this was also necessary, since we had very recently a bug which had been caught, if we had made sure that the html used to test js, had been html rendered in php (there was a change in a template, leading js to fail). It might be good practice to do this, whenever we use example html to test js code upon.
  - I added a one liner to .gitIgnore `/node_modules/**/.cache`, I always have lots of changes in those, when running npm test.
  - Another [one liner](https://github.com/ILIAS-eLearning/ILIAS/compare/trunk...Amstutz:improvement/trunk/js_unit_tests_mocha#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) in package.json, pointing to the npm test to the proper config.

There are some things we should further do soon. I would not like to include them here, since this is big enough already. They are:
- ~~I do not like the way bundling is currently shown in the docs by using the „es“ format. It tends to pollute the global scope of the application. Maybe we should move to iife, however afaik, this makes it hard to expose the public interface to other components with il.UI~~. --> See the discussion with @nhaagen bellow. However, we need to make sure, to initialize il, before loading the UI.
- We should bundle all UI Components js into one single file. This is something, we already agreed upon but I think the time is here to do it.
- We should make further use of the redux-like model proposed in the Mainbar. In the new component I work with it, it makes js a lot easier to test. 
- Finally and most importantly, we should require unit tests for all new js code (at least in the UI Components). I am sure it improves the modularization and overall quality of the JS code a lot. At least it already did IMO for Counter and Notification items.

mBeym committed:
Add new --list option that lists all named objectives

- Add info about new --list command to README.
- Move achieve objective code to separate function

daniwe4 committed:
Study Programme: PHP 8 changes

mBeym committed:
Implement a way for components to listen for user field attribute changes

- Fix issue with second_email => https://mantis.ilias.de/view.php?id=31026
- Add unit tests for user field change listener classes in Services/User/classes

Amstutz committed:
JS Unit tests, some proposals and refactoring towards more and better js unit tests

Amstutz commented:
@mjansenDatabay and @nhaagen I added you two as reviewers as well since you have contributed to unit js unit tests or have experience/proposed aspects of bundling js. If you do not find time to look at it, all good. I understand.

@chfsx , @tfamula and @alex40724 : This might be of interest to you as well, since you all contributed to js in the recent past. Opinions/objections/feedback/questions are always welcome.

nhaagen commented on src/UI/templates/js/Item/rollup.config.js:
```diff
@@ -0,0 +1,7 @@
+  input: './src/notification.js',
+  output: {
+    file: './dist/notification.js',
+    format: 'es'
```
format: 'iife'

nhaagen commented:
Hey @Amstutz ,
thank you very much for this, the location of the tests and linking files really bugged me, but then I simply did not find a proper way to move them. If the "--recursive --config tests/UI/..." does the trick -  hooray, I'm all for it. 
One minor note: if you render the js in IIFE instead of ES, the top namespace will be less cluttered. Give it a try ;)
Otherwise: thumps up, and thanks again.

Amstutz committed:
JS Unit tests, some proposals and refactoring fix rollup

Amstutz committed:
JS Unit tests, some proposals and refactoring fix rollup II

Amstutz commented:
Hi @nhaagen 

Thx very much for the feedback. Highly appreciated. I tested before with iife, there is one (up to this point in time) minor issue, we will run into, we will depend on some external piece of code to globally initialize il for us, right? So this piece of code does not make sense anymore:

```
var il = il || {};
```

Because with IIFE we would only define il internally, if not initialized yet. I do not think this is an issue. But at some point, we will need to make sure, il is initialized before we load UI. However, the benefit is bigger than the risk. I agree. I changed the code and also updated the docs, proposing to make iife the default.

Thx.

alex40724 commented:
1. Bundling

Edit: After reading through all the changes I have the feeling that there is some kind of misinterpretation with the bundling concept.

The bundling.md file currently promotes ES modules. "The single files SHOULD contain export ES6 modules and the overall package SHOULD also be an ES6 module." The change in this file (using "iife" instead) would be strange.

I do not understand how ES modules pollute the top namespace. Is this an UI framework related or a general issue? It should e.g. be possible to bundle the whole UI framework js code in one module which imports the single components (like the component1.js part of the bundling example). Everything would be self contained. If this is currently not possible, we still should not change the general bundling guideline imo.

2. Testing

Using `mocha.config.js` is a good idea, but `tests/UI/Client/mocha.config.js` is a strange place for a global config being referenced in `package.json` imo. The `docs` tests do not belong to the UI framework and we want tests in `Services` and `Modules` also being included. I would suggest to include paths to `./Services/**/test/*.js` and `./Modules/**/test/*.js`.

The `mocha.config.js` could e.g. go directly to the test directory `tests/mocha.config.js` imo.

Amstutz commented:
Hi @alex40724 

Thx a lot for the Feedback and for your time looking into this! I am no expert on this, so take everything with a grain of salt. Thx for helping forstering the discussion here.

1. Bundling

> Edit: After reading through all the changes I have the feeling that there is some kind of misinterpretation with the bundling concept.

That is possible. Therefore, I have no strong opinion as long as we try to be as consistent as possible. 

> "The single files SHOULD contain export ES6 modules and the overall package SHOULD also be an ES6 module." The change in this file (using "iife" instead) would be strange.

Yeah, that would need to be consistent, I will improve this further as soon as we have a decision on the way forwart with bundling.
 
Scoping in ILIAS is performed by putting everything into the il scope. From there we devide further into e.g. il.UI, or il.Foo etc. This prevents as from having naming naming collisions or from having variables, with a scope much to large. We could discuss improving this, but this would probably involve more changes we could handle atm. 

Further I think it is good and common practise to export functionality of modules to other modules using to "export" statement, like so:

```
var X = function(){
...
}

export {X};
```

The problem I did run into, that it was only possible for me to export named functions, e.g. the following is not possible afaik:

```
function(il){
  var X = function(){
  ...
  }
  
  il.X = X;
}(il);

export {il.X};
```

So I did the first option. Leaving X in the global namespace, where it does not belong as pointed out by @nhaagen. So I followed his recommendation and changed to iife, which wraps everything inside a self executing function after compiling. Leaving us with the issue of when we initially define `il`.

> Is this an UI framework related or a general issue?

I do not think so. To me, this will be an issue, wherever we have multiple modules tested individually being exported individually. However, I am sure there coud be other, better solution in place or possible. Do you have some productive JS using ES which you think could be a good modle of how we bundle/modularize/export/import/scope and in future also unit test our js code? Then we can compare the options possible.

> It should e.g. be possible to bundle the whole UI framework js code in one module which imports the single components (like the component1.js part of the bundling example). Everything would be self contained.

Yes this is the plan and will be the immediate next step, once we reach a decision here. See in the initial description, last paragraph:

> We should bundle all UI Components js into one single file. This is something, we already agreed upon but I think the time is here to do it.

So, TL;DR: I am open to other options, as long as we have a pattern that a) does not pollute to global name space and b) makes it easy/possible to unit test small parts of large components.

2. Testing

> Using mocha.config.js is a good idea, but tests/UI/Client/mocha.config.js is a strange place for a global config being referenced in package.json imo.

+1, I definitely see that. This was just in lack of a better idea. I want to have it out of the root dir. This dir is just crumbled with things, which gives a very messy bad first impression. `tests/mocha.config.js` works for me.

> I would suggest to include paths to `./Services/**/test/*.js` and `./Modules/**/test/*.js`.

+1, definitely agree as soon as we have js tests there. Not sure, if there are only the docs and the UI Components featuring such. If there are, we should immediately add the directories. If there are none, I wonder, whether it is not a bid misleading adding those dir at this moment. I do not have strong opinion here. 

Thx for the feedback, looking forward to more discussion on this.

nicoroeser committed:
Remove unnecessary file

Has been introduced in 9fdcbe7aff2cf3112e48b890f28dc196dbc9059e, but
wrongly named (file name contains a space character). An updated version
has been added with the proper name in
1f5542663f4808ecd6cbf609aaabf80685867646. This makes the former file
obsolete, and it should be purged from the tree.

mjansenDatabay committed:
ContentPage: Change default command and return URL when editing page content

(cherry picked from commit 0e041a5d62219a45f1674834d6116959d4554309)

Commit:
Merge pull request #3722 from utesche/patch-3

update repository pattern: ilGeoLocationDBRepository.php

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
Merge pull request #3614 from DatabayAG/trunk_forum_php8

Modifies forum for PHP 8

mjansenDatabay committed:
Forum: Customize Page

Commit:
Global Template, double head Action, see #31156

mjansenDatabay committed:
Forum: Change form input due to changed API in Services/Form (PHP 8)

mjansenDatabay committed:
Mail: Change form input due to changed API in Services/Form (PHP 8)

smeyer-ilias committed:
dtpl: return types

smeyer-ilias committed:
dtpl: return types

smeyer-ilias committed:
dtpl: --GET

smeyer-ilias committed:
log: switched to monolog 2.3.5 (including php 8 support)

smeyer-ilias committed:
log: compatability to monolog 2.3.5

mjansenDatabay committed:
Mail: Fix missing user signature when selecting manual text template (Mantis 31562)

mjansenDatabay committed:
Contacts: PSR-2

M-Barz committed:
adt: refactored dic calls

M-Barz committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

thibsy committed:
added plugin structure support, started writing some tests and changed ilCtrl initialization.

smeyer-ilias committed:
dtpl: fixed settting constructor

klees committed:
#31476: Make User Defined Tags Work in Blog (#3705)

* UI: fix docu of Tag Input Field

* Blog: allow user defined tags for postings

* UI: make user created tags in Tag Input Field work

The implementation of user created tags did not work for various reasons, these
changes ought to fix the problems that where introduced with the switch to tagify.

nhaagen committed:
UI, Mainbar: 30644 render menuitems as list, focus contents (#3725)

alex40724 committed:
php8 fixes

Commit:
This changes the role-attribute in BulkyLinks for Menu-Items (#3737)

Fix for: https://mantis.ilias.de/view.php?id=31552

fsesterh committed:
added security contributor

chfsx committed:
Merge branch 'trunk' into feature/8/clean-temp-folder-in-data-directory-via-cron

# Conflicts:
#	lang/ilias_de.lang
#	lang/ilias_en.lang

chfsx committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into feature/8/clean-temp-folder-in-data-directory-via-cron

chfsx committed:
fixed return value

klees committed:
BackgroundTasks: removed annotation in test which PHPUnit thinks is wrong.

iszmais committed:
Feature/button/minimize implementation (#3664)

* Add minimize button

* fix formattin
g

* fix namespace

* Add context exclusion

* add return type

* Add minimize button implementation and design

* Cleanup delos css

* add lang var

* Add tests

* remove artifacts

* Revert core change

iszmais committed:
Update trunk

iszmais committed:
Update trunk

Amstutz committed:
UI Components, fix of #31529

iszmais committed:
Fix NewMainbar availability calling (#3741)

smeyer-ilias committed:
dtpl: --request

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
Merge pull request #3665 from iszmais/feature/OSC-minimize-button

Feature/osc minimize button

mjansenDatabay committed:
Forum: Fix missing moderator info

mjansenDatabay committed:
Forum: Customize Page

matthiaskunkel committed:
Fixed Mantis #31479 - several typos in German labels

mjansenDatabay committed:
Forum: Customize Page

mjansenDatabay committed:
Merge pull request #3746 from mjansenDatabay/feature/8/frm-copage

ILIAS 8 / Forum: Customize Page

klees commented on docs/development/js/js-unit-test.md:
````diff
@@ -14,18 +15,26 @@ ILIAS is using [Mocha](https://mochajs.org/) as test framework for Javascript un
 }
 ```
 
+Note that it might be good practise to keep the tests cloze to the source an only symlink the files to the above
````
```suggestion
Note that it might be good practise to keep the tests close to the source an only symlink the files to the above
```

M-Barz committed:
DidacticTemplate: Refactored DIC calls (#3747)

DidacticTemplate: Refactored DIC calls

klees reviewed (approved):
Hi @Amstutz,

thanks a lot for iterating on this and cleaning up stuff. I don't feel I have much to add, besides one typo (which I commented on inline).

Regarding `iife` I tried to come up with an opinion by reading a little. As far as I understand, the affairs are as such:

* We ask people to write ES6 modules to modularize their stuff, which gives a clear path to export and import functionality.
* Current state of the art is that components import functionality from other components via `il` and its "namespaces". They do not import functionality from other components via ES6 modules (at least I'm not aware of this).
* Futhermore, current state of the art is that components (such as the ones edited here) use ES6 modules to modularize internally, and then use rollup to bundle their internal modules into one file.
* We ask people to bundle the js of their component as an ES6 module as well.
* Using ES6 modules for a components js (as we currently do) could be another way to import functionality from other, but currently `il` is the way we use.
* @Amstutz changes try to fix this by breaking the ES6 pattern for the bundling of a components js by adding a top level file that shoves the functionality to a sub namespace of `il` and bundling this as `iife` to minimize surprises from changes in the global namespace.

So basically, this seems to be a matter of interoperability between the old `il`-based modularization and the new ES6 based modularization. We would keep using the old model for export/import functionality among components but use ES6 modules in these components internally.

From my perspective this feels plausible. I cannot see how we would switch to an ES6 based model for inter component modularization currently. I would be happy to be told differently, though. So we would keep using the `il` model while improving components internally to use ES6. Some day, when components are in an according shape, the switch to use ES6 throughout the system would become a lot smaller and we could get rid of `il`.

Is this something we want to discuss on Jour Fixe, though?

Best regards!

Amstutz committed:
Update docs/development/js/js-unit-test.md

Co-authored-by: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz committed:
Update docs/development/js/js-unit-test.md

Co-authored-by: Richard Klees <richard.klees@concepts-and-training.de>

Amstutz commented:
Thx for the feedback. Especially for your opinion on the handling of modules. Yes, JF might be a good point to bring this up. This will have some impact for a somewhat extended group of devs.

Amstutz committed:
CSS/Less: Work on Contrast between links and regular text, see: #31099

thibsy committed:
fixed structure artifact bugs and implemented tests for setup

klees committed:
Services/Migration: removed outdated migrations and added roadmap

mjansenDatabay committed:
Chatroom: Cast ID when using object factory

mjansenDatabay committed:
Forum: Fix typo in constructor

mjansenDatabay committed:
Chatroom: PHP 8

mjansenDatabay committed:
Chatroom: PHP 8

smeyer-ilias committed:
adt: replaced sizeof calls

alex40724 committed:
php8 fixes; mostly types

smeyer-ilias committed:
tree: removed constant IL_LAST_NODE in favour of ilTree::POS_LAST_NODE

nhaagen committed:
PHP8: some fixes for issues preventing the creation of new objects

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
dtpl: reformat code

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
tree: db access;

smeyer-ilias committed:
tree: reafctored logging;

mjansenDatabay committed:
UIComponent: Fix return type of ilUIHookPluginGUI::getPluginObject

mjansenDatabay committed:
UIComponent: Fix types in ilButton

mjansenDatabay committed:
Search: Fix basic PHP 8 issues

mjansenDatabay committed:
User: Cast ids to int in ilObjUser::_lookupId

mjansenDatabay committed:
Mail: PHP 8

mjansenDatabay committed:
Contact: PHP 8

mjansenDatabay committed:
RBAC: Cast role ids in \ilRbacReview::getAssignableChildRoles and fix return type

mjansenDatabay committed:
UIComponents/ConfirmationGUI: Fix type issues with accessing variables before assigning a value / Throw exceptions if object state is incomplete on rendering

mjansenDatabay committed:
User: PHP 8 for User/Avatar

mjansenDatabay committed:
User: PHP 8 for User/Password

mjansenDatabay committed:
UIComponents/ConfirmationGUI: Fix type issues with accessing variables before assigning a value / Throw exceptions if object state is incomplete on rendering

Amstutz committed:
A11y, fix of: #31587

mjansenDatabay committed:
Mail: Fix unit tests

mjansenDatabay committed:
User: Cast ids to int in ilObjUser::getUserIdsByInactivityPeriod and ::getUserIdsNeverLoggedIn

mjansenDatabay committed:
User: PHP 8 for User/Gallery

smeyer-ilias committed:
tree: fixed unit test

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
User: PHP 8 ilUserException

nhaagen committed:
JS, AdvancedSelectionList: #31382 add initialization later

mjansenDatabay committed:
RBAC: Cast role ids to int in ilRbacReview::getRolesOfRoleFolder

mjansenDatabay committed:
User: PHP 8 ilCronDeleteNeverLoggedInUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteNeverLoggedInUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

nhaagen committed:
JS, Forms: move JS to global template

alex40724 committed:
Merge pull request #3755 from nhaagen/_trunk/js_forms

JS, Forms: move JS to global template

mjansenDatabay committed:
Utilities: Check array key before accessing it (ilMWFakery)

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
UIComponent: Fix 'Access key function ids are integers not strings'

alex40724 committed:
Merge pull request #3756 from mjansenDatabay/hotfix/8/access-keys

UIComponent: Fix 'Access key function ids are integers not strings'

alex40724 committed:
fixed wrong type

alex40724 committed:
Merge pull request #3753 from nhaagen/_trunk/0031382

JS, AdvancedSelectionList: #31382 add initialization later

nhaagen committed:
JS, Forms: 'softer' use of  global; fix broken tests

smeyer-ilias committed:
Merge pull request #3667 from mjansenDatabay/improvement/8/remove-debug-functions

Remove unused debug file with debug functions

mjansenDatabay committed:
Init/SessionHandler: Fix review remarks regarding SessionHandlerInterface

smeyer-ilias committed:
Merge pull request #3656 from mjansenDatabay/improvement/8/sessions-php-version-conditionasl

Session Handler: Implement interface / Remove PHP version conditional…

mjansenDatabay committed:
Cron: PHP 8 after types have been changed in legacy forms

mjansenDatabay committed:
UIComponent/ilMultiSelectInputGUI: Handle missing parameter in post body when ilMultiSelectInputGUI::setValueByArray is called

mjansenDatabay committed:
User: PHP 8 ilCronDeleteNeverLoggedInUserAccounts

thibsy committed:
implemented further ilCtrl tests and fixed some small bugs.

alex40724 committed:
php8 fixes

alex40724 committed:
added tests for request class

mjansenDatabay committed:
LDAP: ilLDAPRoleAssignmentRules::getAttributeNames PHP 8 (partly) + PSR-2

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

matthiaskunkel committed:
Improved labels for settings in booking tool

mjansenDatabay committed:
Tree: Fix wrong tree root id fallback in constructor

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactivatedUserAccounts

mjansenDatabay committed:
User: PHP 8 ilCronDeleteInactiveUserAccounts

mjansenDatabay committed:
Forum: Use refinery for cron job configuration

mjansenDatabay committed:
Mail: Use refinery for cron job configuration

mjansenDatabay committed:
Forum: Make save interval mandatory if auto save is enabled for drafts

mjansenDatabay committed:
Forum: Make save interval mandatory if auto save is enabled for drafts

mjansenDatabay committed:
Chat: PHP 8

mjansenDatabay committed:
SAML: PHP 8

mjansenDatabay committed:
ToS: PHP 8

mjansenDatabay committed:
Mail: PHP 8

mjansenDatabay committed:
Contact: PHP 8

mjansenDatabay committed:
Forum: PHP 8

mjansenDatabay committed:
Forum: PHP 8

mjansenDatabay committed:
Chat: PHP 8.1 (Upload)

mjansenDatabay committed:
Certificate: PHP 8.1 (Upload)

mjansenDatabay committed:
Chat: PHP 8.1 (Upload)

mjansenDatabay committed:
CmiX: PHP 8.1 (Upload)

mjansenDatabay committed:
LTI: PHP 8.1 (Upload)

mjansenDatabay committed:
A11Y: PHP 8.1 (Upload)

mjansenDatabay committed:
WorkflowEngine: PHP 8.1 (Upload)

mjansenDatabay committed:
WebDAV: PHP 8.1 (Upload)

mjansenDatabay committed:
User: PHP 8.1 (Upload)

mjansenDatabay committed:
ToS: PHP 8.1 (Upload)

mjansenDatabay committed:
OpenIDConnect: PHP 8.1 (Upload) + PSR-2

mjansenDatabay committed:
Object: PHP 8.1 (Upload)

mjansenDatabay committed:
Chat: PHP 8

mjansenDatabay committed:
Chat: PHP 8

mjansenDatabay committed:
Chat: PHP 8

Amstutz committed:
UI Components, System Notifications, Fix of #31391

mjansenDatabay committed:
Object: Fix form title/icon in creation screen (PHP 8 types)

mjansenDatabay committed:
Merge pull request #3763 from mjansenDatabay/hotfix/8/31026/mail

Implement a way for components to listen for user field attribute cha…

chfsx committed:
fixed setTarget in form

chfsx committed:
update steps

chfsx committed:
[FIX] 0031185: IRSS: refactoring of DB structures

# Conflicts:
#	Modules/File/classes/class.ilObjFileAccess.php
#	Modules/File/classes/class.ilObjFileListGUI.php
#	setup/sql/7_hotfixes.php

mjansenDatabay committed:
Mail/User: Apply tiny code style and documentation changes in addition to fix for Mantis issue 31026

alex40724 committed:
php8 fixes, mostly types and missing declarations

alex40724 committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
Mail/User: Apply tiny code style and documentation changes in addition to fix for Mantis issue 31026

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
Merge pull request #3761 from mjansenDatabay/hotfix/8/tree-root-id

Tree: Fix wrong tree root id fallback in constructor

chfsx committed:
Merge branch 'improvement/8/irss-db-structures' into trunk

thibsy committed:
fixed ILIAS tests and added further ilCtrl tests.

thibsy committed:
merge trunk into proposal/8/static-ilctrl

thibsy committed:
fixed merge of trunk into proposal/8/static-ilctrl

thibsy committed:
resolved conflicts

thibsy committed:
removed merge-confligt statements from ilCharSelectorGUI

thibsy committed:
fixed context path progression

thibsy committed:
fixed ilCtrlContext path adoption

thibsy committed:
removed Services/Feeds/classes/class.ilPDExternalFeedBlockGUI.php

thibsy committed:
added (almost) empty test directory

thibsy committed:
removed ilCtrlIterator test exception messages, because they differ in PHP versions.

matthiaskunkel commented:
Jour Fixe, 15 NOV 2021 :  Timon made a short presentation of his PR and its current discussion. We do not have any objections against using iife for creating the bundles. The mocha.config.js should go directly to the test directory tests/mocha.config.js. Please merge PR to trunk accordingly.

alex40724 committed:
php8: request handling

chfsx committed:
Fixed ilCtrl setCmd with empty commands (for "return" in repository and other places)

smeyer-ilias committed:
grp: small fixes

smeyer-ilias committed:
small case typo

smeyer-ilias committed:
tree: return types

smeyer-ilias committed:
tree: return types ilPathGUI::getPath();

alex40724 committed:
added tests

M-Barz committed:
BT-31484: Fixed activation of subtabs in group settings (#3767)

alex40724 committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

alex40724 committed:
php8: added some types

smeyer-ilias committed:
0031361: Metadata: Selected language gets lost during copy process, other metadata is copied

Commit:
ilDBUpdate:: values in [settings] to be changed MUST be strictly of type 'string' (#3773)

(v8.0alpha)
Otherwise type errors during 'setup.php install' aborts processing of the install command.

Commit:
Docs: some typos

thibsy committed:
moved CSRF token storage to ilSession and adjusted tests accordingly.

thibsy committed:
implemented filter method in structure-helper that removes unnecessary structure entries.

smeyer-ilias committed:
0031000: Feldtyp Auswahlliste (Einfachauswahl) und Auswahlliste (Mehrfachauswahl) funktionieren nicht im Portfolio

thibsy committed:
fixed command retrieval, adjusted tests to pass

Amstutz committed:
System Styles, Icon Color, fix of #31625, #31624, #31499

Amstutz committed:
UI Components, fix of #30314, #31626

matthiaskunkel committed:
Improved labels for anonymous tests in agreement with 1st maintainer (thanks to Denis)

mjansenDatabay committed:
Chatroom: PHP 8

mjansenDatabay committed:
APC: Fix PHP type when retrieving memory setting (#3757)

M-Barz committed:
Merge pull request #3713 from klees/release_7_31213

ErrorHandling: only show error message once #31213

mBeym committed:
Fix objective not created when no preconditions are defined

mBeym committed:
Remove unnecessary else

mBeym committed:
Add test for getAgents function

mBeym committed:
Add test for achieving objective

mBeym committed:
Add test for achieving invalid objective

mBeym committed:
Add tests for ObjectiveConstructor

mBeym committed:
Add tests for currenty available named objectives

mBeym committed:
Fix failing test

mBeym committed:
Apply php-cs-fixer

smeyer-ilias committed:
tree: return types ilTreeImplementation; ensure ROOT_FOLDER_ID int

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
0030665: Column "Organisationeinheiten" empty in participant lists

smeyer-ilias committed:
tree: interface return types

klees committed:
Setup: minor formatting

klees committed:
Setup: removed unneccessary specificity in objective constructor

klees committed:
Setup: minor naming improvement in test helper

klees committed:
Setup: fixed tests

klees committed:
Setup/UICore: moved test to correct component

klees committed:
Setup: fix config passing in AgentCollection::getNamedObjectives

klees committed:
Setup: correct tests for AchieveCommand

klees committed:
Setup: minor doc change in AchieveCommand

padvincenzo committed:
Sort icons and remove unnecessary full path from Edit icons (#3784)

* Added public function sortIconsByPath()

* Sort icons and remove unnecessary full path from Edit icons

* Update Services/Style/System/classes/Icons/class.ilSystemStyleIconFolder.php

Co-authored-by: Michael Jansen <mjansen@databay.de>

* Update Services/Style/System/classes/Icons/class.ilSystemStyleIconFolder.php

Co-authored-by: Michael Jansen <mjansen@databay.de>

* Update Services/Style/System/classes/Icons/class.ilSystemStyleIconFolder.php

Co-authored-by: Michael Jansen <mjansen@databay.de>

Co-authored-by: Michael Jansen <mjansen@databay.de>

Amstutz committed:
CSS, Fix Contrast issue with Text and BG for a11y

M-Barz committed:
sysc: PHP 8 changes - return and param types, globals, quotes, clean-up (#3759)

* sysc: PHP 8 changes - return and param types, globals, quotes, clean-up

* sysc: Moved direct $_REQUEST access to wrapper in ilSCTreeTasksGUI

* sysc: added missing return types

* sysc: reformatted code

* sysc: added visibility for constants

M-Barz committed:
sysc: fixed warnings from code inspection

smeyer-ilias committed:
tree: ilPathGUI type hints

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

mjansenDatabay committed:
ContentPage: Add more tests

mjansenDatabay committed:
Chatroom: Fix 'any' expectation issues in tests

mjansenDatabay committed:
Object/CustomIcon: PHP 8

mjansenDatabay committed:
Object/CustomIcon: PHP 8

mjansenDatabay committed:
Object/ObjectService: PHP 8

mjansenDatabay committed:
Object/CommonSettings: PHP 8

thibsy committed:
added further dependency injection and some minor renaming.

thibsy committed:
Merge branch 'trunk' into proposal/8/static-ilctrl

lscharmer committed:
Refinery: MakeClickable Transformation (#3782)

* Add MakeClickable transformation

mjansenDatabay committed:
Refinery: PSR-2

thibsy committed:
fixed ilUICoreSetupAgent

thibsy committed:
Merge pull request #3710 from studer-raimann/proposal/8/static-ilctrl

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
0031633: Textfields do not work in Copied Local ADT (was: Metadata Text Field in Portfolio not Working!)

chfsx committed:
Merge branch 'trunk' into feature/8/clean-temp-folder-in-data-directory-via-cron

chfsx committed:
correct comment

chfsx committed:
Merge branch 'trunk' into feature/8/clean-temp-folder-in-data-directory-via-cron

# Conflicts:
#	lang/ilias_de.lang
#	lang/ilias_en.lang

chfsx committed:
fixed en variable for new cronjob

mjansenDatabay committed:
ContentPage: Add more tests

lscharmer committed:
E-Mail: clickable mixed content type (#3779)

* Add new Transformation to make a text clickable (with HTML anchors)

smeyer-ilias committed:
tree: php 8 return types, test

smeyer-ilias committed:
log: avoid static call

mjansenDatabay committed:
Contacts: Got rid of static method calls

mjansenDatabay committed:
Merge branch 'trunk' of github.com:ILIAS-eLearning/ILIAS into trunk

rsheer committed:
Fix import of data collections if certain values are not set (#3574)

@rsheer Thanks a lot! And sorry for the late answer.

klees committed:
Setup: list objectives for achieve if no objective is defined

klees committed:
Setup: fix objective listing for achieve

mjansenDatabay committed:
Contacts: Add more unit tests

thibsy committed:
fixed wrong ilCtrl iterator in ilPluginDBUpdate.

Amstutz committed:
Merge branch 'improvement/trunk/js_unit_tests_mocha' of https://github.com/Amstutz/ILIAS into improvement/trunk/js_unit_tests_mocha

Conflicts:
	tests/UI/Client/Input/Container/filter.test.js
	tests/UI/Client/MainControls/mainbar.test.js
	tests/UI/Client/Menu/Drilldown/drilldown.test.js

Amstutz committed:
UI Components Fixing Tests

Amstutz committed:
UI Components, moving mocha.config.js

Amstutz commented:
I added the changes as requested by jf. However, I will manually pick the changes to trunk. I seem to have messed something up while rebasing my changes onto trunk.
